pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -985,17 +959,60 @@ public void addTable(AbstractTableConfig config) throws JsonGenerationException,
+  private void ensureRealtimeClusterIsSetUp(AbstractTableConfig config, String realtimeTableName,
+      IndexingConfig indexingConfig) {
+    KafkaStreamMetadata kafkaStreamMetadata = new KafkaStreamMetadata(indexingConfig.getStreamConfigs());
+    IdealState idealState = _helixAdmin.getResourceIdealState(_helixClusterName, realtimeTableName);
+
+    if (kafkaStreamMetadata.hasHighLevelKafkaConsumerType()) {
+     if (kafkaStreamMetadata.hasSimpleKafkaConsumerType()) {
+       // We may be adding on low-level, or creating both.
+       if (idealState == null) {
+         // Need to create both. Create high-level consumer first.
+         createHelixEntriesForHighLevelConsumer(config, realtimeTableName, idealState);
+         idealState = _helixAdmin.getResourceIdealState(_helixClusterName, realtimeTableName);
+         LOGGER.info("Configured new HLC for table {}", realtimeTableName);
+       }
+       // Fall through to create low-level consumers
+     } else {
+       // Only high-level consumer specified in the config.
+       createHelixEntriesForHighLevelConsumer(config, realtimeTableName, idealState);
+     }
+    }
+
+    // Either we have only low-level consumer, or both.
+    if (kafkaStreamMetadata.hasSimpleKafkaConsumerType()) {
+      // Will either create idealstate entry, or update the IS entry with new segments
+      // (unless there are low-level segments already present)
+      final String llcKafkaPartitionAssignmentPath = ZKMetadataProvider.constructPropertyStorePathForKafkaPartitions(realtimeTableName);
+      if(_propertyStore.exists(llcKafkaPartitionAssignmentPath, AccessOption.PERSISTENT)) {
+        PinotTableIdealStateBuilder.buildLowLevelRealtimeIdealStateFor(realtimeTableName, config, _helixAdmin,
+            _helixClusterName, idealState);
+        LOGGER.info("Successfully added Helix entries for low-level consumers for {} ", realtimeTableName);
+      } else {
+        LOGGER.info("LLC is already set up for table {}, not configuring again", realtimeTableName);
+      }
+    }
+  }
+
+  private void createHelixEntriesForHighLevelConsumer(AbstractTableConfig config, String realtimeTableName,
+      IdealState idealState) {
+    if (idealState == null) {
+      idealState = PinotTableIdealStateBuilder
+          .buildInitialHighLevelRealtimeIdealStateFor(realtimeTableName, config, _helixAdmin, _helixClusterName,
+              _propertyStore);
+      LOGGER.info("Adding helix resource with empty HLC IdealState for {}", realtimeTableName);
+      _helixAdmin.addResource(_helixClusterName, realtimeTableName, idealState);
+    } else {
+      // TODO jfim: We get in this block if we're trying to add a HLC or it already exists. If it doesn't already exist, we need to set instance configs properly (which is done in buildInitialHighLevelRealtimeIdealState, surprisingly enough). For now, do nothing.
+      LOGGER.info("Not reconfiguring HLC for table {}", realtimeTableName);
```

Comment:
Can't really throw an exception, since this is going to get called every time we update the indexing config (eg. to change metadata) and an HLC config already exists. This would prevent metadata updates and/or other indexing config changes.